File: 백준/Gold/4195. 친구 네트워크/친구 네트워크.py
import sys
input = sys.stdin.readline



def find_parent(x):
    if parent[x] ==x:
        return x
    parent[x] = find_parent(parent[x])
    return parent[x]

def union_parent(a,b):
    a = find_parent(a)
    b = find_parent(b)
    pre_sum = cnt[a]+cnt[b]
    if a>b:
        parent[a] = b
        cnt[a] = pre_sum
        cnt[b] = cnt[a]
    elif a<b:
        parent[b] = a
        cnt[b] = pre_sum
        cnt[a] = cnt[b]


for _ in range(int(input())):
    n = int(input())
    idx = 0
    cnt = dict()
    parent = dict()
    for i in range(n):
        user1, user2 = map(str,input().rstrip().split())
        
        if user1 not in parent:
            parent[user1] = user1
            cnt[user1] = 1
        if user2 not in parent:
            parent[user2] = user2
            cnt[user2] = 1
        
        union_parent(user1, user2)

        print(cnt[find_parent(user1)])

File: 백준/Gold/4195. 친구 네트워크/친구 네트워크.rs
use std::collections::HashMap;
use std::io::{self, BufRead};

fn find(parent: &mut HashMap<String, String>, x: &str) -> String {
    let parent_x = parent[x].clone();
    if parent_x != x {
        let root = find(parent, &parent_x);
        parent.insert(x.to_string(), root.clone());
        root
    } else {
        x.to_string()
    }
}

fn union(parent: &mut HashMap<String, String>, cnt: &mut HashMap<String, i32>, a: &str, b: &str) {
    let root_a = find(parent, a);
    let root_b = find(parent, b);
    if root_a != root_b {
        let pre_sum = cnt[&root_a] + cnt[&root_b];
        if root_a > root_b {
            parent.insert(root_a.clone(), root_b.clone());
            cnt.insert(root_b.clone(), pre_sum);
        } else {
            parent.insert(root_b.clone(), root_a.clone());
            cnt.insert(root_a.clone(), pre_sum);
        }
    }
}

fn main() {
    let stdin = io::stdin();
    let input = stdin.lock().lines();
    let mut iter = input.map(|line| line.unwrap());

    let t: usize = iter.next().unwrap().parse().unwrap();
    let mut sb = String::new();

    for _ in 0..t {
        let n: usize = iter.next().unwrap().parse().unwrap();
        let mut parent: HashMap<String, String> = HashMap::new();
        let mut cnt: HashMap<String, i32> = HashMap::new();

        for _ in 0..n {
            let line = iter.next().unwrap();
            let mut users = line.split_whitespace();
            let user1 = users.next().unwrap().to_string();
            let user2 = users.next().unwrap().to_string();

            if !parent.contains_key(&user1) {
                parent.insert(user1.clone(), user1.clone());
                cnt.insert(user1.clone(), 1);
            }
            if !parent.contains_key(&user2) {
                parent.insert(user2.clone(), user2.clone());
                cnt.insert(user2.clone(), 1);
            }
            union(&mut parent, &mut cnt, &user1, &user2);
            sb.push_str(&cnt[&find(&mut parent, &user1)].to_string());
            sb.push_str("\n");
        }
    }

    print!("{}", sb);
}